### PR TITLE
Add Dockerfile to build docker image from ScanCode sources

### LINKED ISSUE #2264: Add Dockerfile to support building Docker images
## Short Description
Add a Dockerfile to allow building Docker images for ScanCode Toolkit, to avoid dependency management issues and enable a simple workflow to get started.

## Possible Labels
- installation and packaging

## Select Category
-  [ ] Enhancement
-  [ ] Add License/Copyright
-  [ ] Scan Feature
-  [x] Packaging
-  [x] Documentation
-  [ ] Expand Support
-  [ ] Other

## **How This Feature will help you/your organization**
Allows using the same workflow independent of the platform, avoids potential issues with dependency management. 

## **Can you help with this Feature**
A pull request will be provided presently.

### PR DESCRIPTION
**Pull request by daniel-eder**
Fixes #2264

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Ignore the Docker files themselves to avoid that all layers are invalidated if the Docker files are changed during development.
+.dockerignore
```
Comment (pombredanne) [inline .dockerignore:2]: You may want to also ignore the samples, doc and tests directories?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+RUN mkdir scancode-toolkit
+
+# Copy sources into docker container
+COPY . scancode-toolkit
+
+# Set workdir
+WORKDIR scancode-toolkit
+
+# Run scancode once for configuration
+RUN ./scancode --help
```
Comment (pombredanne) [inline Dockerfile:40]: You should instead run `--reindex-licenses` IMHO to create the base license index

File: docs/source/getting-started/install.rst [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 
 - :ref:`app_install`
 
     The recommended method for installing ScanCode is Downloading the latest release as an
     application and then configure and use directly. This is easy because no knowledge of pip/git
     or other developer tools is necessary.
 
+- :ref:`docker_install`
+
+    An alternative to installing the latest Scancode Toolkit release natively is to build a Docker image from the included Dockerfile. This is easy because the only prerequisite is a working Docker install.
```
Comment (pombredanne) [inline docs/source/getting-started/install.rst:14]: Do you mind avoiding having too lines that are too long? 90'ish characters should be about right.

File: docs/source/getting-started/install.rst [hunk 4]
```diff
@@ -234,6 +238,45 @@ Un-installation
...
+
+The docker image will forward all arguments it receives directly to the ``scancode`` command. 
+
+Display help:
+
+    docker run scancode-toolkit --help
+
+Mount current working directory and run scan on mounted folder:
+
+    docker run -v $(pwd):/project scancode-toolkit <arguments> /project
```
Comment (pombredanne) [inline docs/source/getting-started/install.rst:274]: And where do go the output?

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you ++! this is awesome!
There a few comments for your review and there are some checks on the doc that are missing.
Do you mind to also your name to the AUTHORS file and adding a CHANGELOG entry?

File: docs/source/getting-started/install.rst (hunk 4 above)
Comment (daniel-eder) [reply docs/source/getting-started/install.rst:274]: I'll add a line explaining and expand the sample - any output written to the mounted directory like `/project/my-output-file.json` inside the docker command will create a file `$PWD/my-output-file.json` in the current working directory on the host machine. Alternatively, a second directory could be mounted with a further -v command to create a separate output directory.

**Comment by pombredanne** [comment]
>  The travis build fail seems not to be due to my changes though:

@AyanSinhaMahapatra would you know why this is happening by chance?

**Comment by AyanSinhaMahapatra** [comment]
@pombredanne and @daniel-eder Really sorry for the late reply, I missed this somehow. 

So the issue is there are stylistic checks done as a part of documentation tests, using doc8, and that's causing the fails as there are some changes in `install.rst`. Docs are here - https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_doc.html#how-to-run-style-tests

These are the errors, you can see if you scroll down. 

```
1.64s$ ./scripts/doc8_style_check.sh
source/getting-started/install.rst:14: D002 Trailing whitespace
source/getting-started/install.rst:15: D002 Trailing whitespace
source/getting-started/install.rst:245: D002 Trailing whitespace
source/getting-started/install.rst:260: D002 Trailing whitespace
source/getting-started/install.rst:269: D002 Trailing whitespace
source/getting-started/install.rst:279: D002 Trailing whitespace
source/getting-started/install.rst:280: D002 Trailing whitespace
source/getting-started/install.rst:281: D002 Trailing whitespace
source/getting-started/install.rst:283: D002 Trailing whitespace
The command "./scripts/doc8_style_check.sh" exited with 1.
```

These below were just warnings and didn't cause the test to fail.

```
/home/travis/build/nexB/scancode-toolkit/docs/source/tutorials/how_to_visualize_scan_results.rst:30: WARNING: undefined label: workbench_requirements (if the link has no caption the label must precede a section header)
/home/travis/build/nexB/scancode-toolkit/docs/source/tutorials/how_to_visualize_scan_results.rst:79: WARNING: undefined label: workbench_views (if the link has no caption the label must precede a section header)
```

Just removing the whitespace at these lines will pass the tests. Sorry for the inconvenience. Thank You for the contributions.

**Comment by pombredanne** [comment]
@AyanSinhaMahapatra I think you forgot to signoff your commit :) No biggie... I did set the DCO to pass

**Comment by pombredanne** [comment]
@AyanSinhaMahapatra you can just add your signoff in a message

**Comment by AyanSinhaMahapatra** [comment]
Update install.rst to pass doc8 tests …05933ed
Remove trailing whitespaces
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

@pombredanne Thanks! All the tests should pass now. Sorry, I edited through GitHub and forgot to signoff.

**Comment by daniel-eder** [comment]
> moving the whitespace at these lines will pass the tests. Sorry for the inconvenience. Thank You for the contributions.
@AyanSinhaMahapatra Thanks a lot for fixing that!

**Review by pombredanne — APPROVED** [review]
All good now! Thanks!

**Pull request merged**